app/src/main/java/fr/free/nrw/commons/nearby/fragments/NearbyParentFragment.java

```diff
@@ -316,15 +319,17 @@ public void onActivityResult(Map<String, Boolean> result) {
                         getActivity().getString(android.R.string.ok),
                         getActivity().getString(android.R.string.cancel),
                         () -> {
-                            if (!(locationManager.isNetworkProviderEnabled()
-                                || locationManager.isGPSProviderEnabled())) {
-                                showLocationOffDialog();
-                            }
+                            askForLocationPermission();
                         },
-                        () -> isPermissionDenied = true,
+                        null,
                         null,
                         false);
                 } else {
+                    if (isPermissionDenied) {
+                        locationPermissionsHelper.showAppSettingsDialog(getActivity(),
+                            R.string.nearby_needs_location);
+                    }
+                    Timber.d("The user checked 'Don't ask again' or denied the permission twice");
                     isPermissionDenied = true;
```

Comment:
Given the existence of the `LocationPermissionsHelper.requestForLocationAccess` method, could we not do away with the `locationPermissionLauncher` code? 🤔 

Let me know in case I missed something.

Reply:
I tried to do that, but the required behaviours of the Nearby and Explore fragment's maps are a bit different from what is implemented in the LocationPermissionsHelper.requestForLocationAccess 
Ex:
1. both these fragments have to load pins/images even when permission is not there, whereas the location picker doesn't do that
2. Even the dialogs flow is a bit different if I remember correctly 
I have used the LocationPermissionsHelper for all other works, such as showing dialogs, opening settings, etc, to reuse code. However, I believe using it for requesting location permission for Nearby and Explore might cause problems.

Reply:
> both these fragments have to load pins/images even when permission is not there, whereas the location picker doesn't do that

I'm confused. We would be using LocationPermissionHelper only to request permission when necessary. How would that actually hinder the usual flow of Nearby? We do have callbacks that could be used to trigger the behaviour when permission is granted / denied. What am I missing?

> Even the dialogs flow is a bit different if I remember correctly

Could you clarify any specific flow that you observe a difference with? I wasn't able to find the difference as far as I checked.

> I have used the LocationPermissionsHelper for all other works, such as showing dialogs, opening settings, etc, to reuse code. 

Indeed. The permission request code alone seems redundant now.

>  However, I believe using it for requesting location permission for Nearby and Explore might cause problems.

If it causes problems, we could certainly live with duplication. I just wanted to ensure we're not missing a good chance to curb redundancy :wink: 

If we end up noticing any discrepancy, it would be worth capturing the same as code comments as it would help for future reference.

Reply:
> I'm confused. We would be using LocationPermissionHelper only to request permission when necessary. How would that actually hinder the usual flow of Nearby? We do have callbacks that could be used to trigger the behaviour when permission is granted / denied. What am I missing?

I had tried to make use of the LocationPermissionHelper's callbacks only, but these were the problems I faced:

The Explore and Nearby use the logic that if permission is not there and the `shouldShowRequestPermissionRationale` method returns false, that means the user has denied permission twice (because this is checked only after permission is asked for once) and thus we cannot ask the user for permission and a dialog needs to be shown to ask the user to give permission from settings, so we set `isPermissionDenied` to true. 

The flow of `LocationPermissionHelper's` `requestForLocationAccess` is such that first, the location permission is checked, and if not given then the `shouldShowRequestPermissionRationale` method is checked. Now, this method returns false in 2 scenarios, one where permission has never been asked for and another where the user had denied permission once. If the method returns false, then we ask for permission.

Thus, we cannot rely on the value of the `shouldShowRequestPermissionRationale` method in `LocationPermissionHelper` (since it will return false even when the permission has never been asked for) to set the value of `isPermissionDenied` in Explore and Nearby fragments.


> > Even the dialogs flow is a bit different if I remember correctly
> 
> Could you clarify any specific flow that you observe a difference with? I wasn't able to find the difference as far as I checked.

In `LocationPickerActivity`, if user denies permission for the first time, then only a toast is shown to them that location permission is denied (so as to not obstruct their picking of location, as they are going to upload an image) and a rationale dialog is shown only on clicking the target button (or leaving the activity and coming back to it). But, in Nearby and Explore, when the user denies permission for the first time, we show a rationale dialog to user, to ask for permission again.

> > I have used the LocationPermissionsHelper for all other works, such as showing dialogs, opening settings, etc, to reuse code.
> 
> Indeed. The permission request code alone seems redundant now.
> 
> > However, I believe using it for requesting location permission for Nearby and Explore might cause problems.
> 
> If it causes problems, we could certainly live with duplication. I just wanted to ensure we're not missing a good chance to curb redundancy 😉
> 
> If we end up noticing any discrepancy, it would be worth capturing the same as code comments as it would help for future reference.

I totally agree with you and thus put in every effort to remove all redundancies present and fix Nearby and Explore's intended behaviour simultaneously. I'll go through the entire logic again and check if the permission redundancy can be removed and raise a separate PR, if you agree, since that might take more time to check all scenarios (and this would delay the release of the new version) . In the meantime, I can add comments regarding why we are not using `LocationPermissionHelper's` `requestForLocationPermission` if you say so :)